command/operator_debug.go

```diff
@@ -876,6 +897,32 @@ func (c *OperatorDebugCommand) collectAgentHost(path, id string, client *api.Cli
 	c.writeJSON(path, "agent-host.json", host, err)
 }
 
+func (c *OperatorDebugCommand) collectPeriodicPprofs(client *api.Client) {
+	duration := time.After(c.duration)
+	// Set interval to 0 so that we immediately execute, wait the interval next time
+	pprofInterval := time.After(0 * time.Second)
+	var pprofIntervalCount int
+	var name string
+
+	for {
+		select {
+		case <-duration:
+			c.cancel()
+			return
+
+		case <-pprofInterval:
+			name = fmt.Sprintf("%04d", pprofIntervalCount)
+			c.Ui.Output(fmt.Sprintf("    Capture pprofInterval %s", name))
+			c.collectPprofs(client)
+			pprofInterval = time.After(c.pprofInterval)
```

Comment:
`time.After` spins up a goroutine and we're doing it in a loop here. While in practice if it's interrupted we'll leak at most one goroutine and then immediately exit the process, it's better to do this cleanly so that folks don't copy-paste it around the codebase.

For this sort of thing I'd use a [`time.Ticker`](https://pkg.go.dev/time#Ticker) with a `defer ticker.Stop()`. Because the `c.collectPprofs` method doesn't take a context, we don't have to worry about it being interrupted. For the first "tick", we can just call the `c.collectPprofs` outside the loop.

Reply:
That's a TIL for me, thanks for pointing it out. I initially confused it with [`time.After() bool`](https://pkg.go.dev/time#Time.After) for some reason but it makes complete sense now going through the source once. Made the relevant changes.